### PR TITLE
Feat/130 expose predefined templates

### LINKED ISSUE #129: Templates for non-default groupStrategy are broken
Both [default.hbs](https://github.com/astahmer/openapi-zod-client/blob/3933722e0c6d7a7b7af50dd63d31aa5f91f666a1/lib/src/templates/default.hbs) and [grouped.hbs](https://github.com/astahmer/openapi-zod-client/blob/3933722e0c6d7a7b7af50dd63d31aa5f91f666a1/lib/src/templates/grouped.hbs) are currently broken when multiple Zodios `Api` objects are defined.

Additionally the grouped template doesn't import the `ZodiosOptions` type, looks like the `createApiClient` was blindly copied from the default template. Probably a good idea to add integration tests which verify the `groupStrategy` functionality, checking for valid typescript with no errors for [petstore-minimal.json](https://github.com/astahmer/openapi-zod-client/blob/3933722e0c6d7a7b7af50dd63d31aa5f91f666a1/examples/petstore-minimal.json) would probably be sufficient.

### PR DESCRIPTION
**Pull request by narakhan**
@astahmer added support for #130, then implemented a template to address #131. While I was at it refactored at the common template code into Handlebar partials, used that to fix #129 and various other issues identified with the templates.

Also added support for the `x-enumName` extension attribute, as otherwise integer enums aren't associated with their names. Additionally I updated the contributing section of the readme and added support for Corepack and a `.node-version` file for whatever Node.js manager people happen to use, as well as bumping the dependencies.

Made sure to update the tests, but didn't add the integration tests mentioned in #129 as I ran out of time. Made sure to do extensive testing against my workplaces' private OpenAPI specification, verified all modified settings.

There's a slight potential for breaking changes with this PR, if a consumer of the library had templates named `schemas-only` or `export-schemas` then it'll cause issues (but that seems unlikely).

### CODE REVIEW

**Comment by narakhan** [comment]
Didn't have time to fix it, but noticed the types generated to handle cyclical data structures are all labeled `Partial` regardless of the `withAlias` variable. The `TemplateContext["options"]` aren't being provided to `getTypescriptFromOpenApi` so that likely needs to be plumbed into it.

File: README.md [hunk 1]
```diff
@@ -285,6 +285,16 @@ NOT tested/expected to work with OpenAPI before v3, please migrate your specs to
...
+```bash
+> pnpm install
+> pnpm test
+```
+
+Assuming no issue were raised by the tests, you may use `pnpm dev` to watch for code changes during development.
+
+If you fix an edge case please make a dedicated minimal reproduction test in the [`tests`](./tests) folder so that it doesn't break in future versions
+
+Make sure to generate a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) before submitting your PR.
```
Comment (astahmer) [inline README.md:300]: thank you for this, was much needed

File: lib/tests/group-strategy.test.ts [hunk 2]
```diff
@@ -565,10 +567,20 @@ test("group-strategy", async () => {
...
       export const putApi = new Zodios(putEndpoints);
 
-      export function createApiClient(baseUrl: string, options?: ZodiosOptions) {
-        return new Zodios(baseUrl, endpoints, options);
+      export function createApiClient(
+        endpointGroup: keyof typeof endpoints,
+        baseUrl: string,
+        options?: ZodiosOptions
+      ) {
+        return new Zodios(baseUrl, endpoints[endpointGroup], options);
```
Comment (astahmer) [inline lib/tests/group-strategy.test.ts:583]: this looks great !

File: lib/src/generateZodClientFromOpenAPI.ts [hunk 3]
```diff
@@ -3,19 +3,21 @@ import path from "node:path";
...
 import { maybePretty } from "./maybePretty";
 import type { TemplateContext } from "./template-context";
 import { getZodClientTemplateContext } from "./template-context";
 
+export const PredefinedTemplates = ["schemas-only", "export-schemas"] as const;
+
 type GenerateZodClientFromOpenApiArgs<TOptions extends TemplateContext["options"] = TemplateContext["options"]> = {
     openApiDoc: OpenAPIObject;
-    templatePath?: string;
+    templatePath?: typeof PredefinedTemplates | string;
```
Comment (astahmer) [inline lib/src/generateZodClientFromOpenAPI.ts:17]: I think it should be 

```ts
templatePath?: typeof PredefinedTemplates[number] | (string & {});
```
Screenshot 2023-05-04 at 10 19 30

File: lib/src/openApiToZod.ts [hunk 4]
```diff
@@ -146,6 +148,15 @@ export function getZodSchema({ schema, ctx, meta: inheritedMeta, options }: Conv
                 return code.assign("z.never()");
             }
 
+            if (schema["x-enumNames"]?.length === schema.enum?.length) {
```
Comment (astahmer) [inline lib/src/openApiToZod.ts:151]: I didn't know about this `x-enumNames`, from what I've found it looks like it's coming from NSwag ? 
could you add a comment with a link to the spec reference so that we know this one is special

Comment (astahmer) [inline lib/src/openApiToZod.ts:151]: also, do you think it would make sense here to include a check on `length === 1` like the others to simplify the output when possible ?

**Review by astahmer — CHANGES_REQUESTED** [review]
you're on fire here, thank you for your work !

**Comment by astahmer** [comment]
> Didn't have time to fix it, but noticed the types generated to handle cyclical data structures are all labeled `Partial` regardless of the `withAlias` variable. The `TemplateContext["options"]` aren't being provided to `getTypescriptFromOpenApi` so that likely needs to be plumbed into it.

got it, unless you want to do it here, it might be worth opening an issue so it can be tracked and hopefully done at some point

File: lib/tests/group-strategy.test.ts [hunk 5]
```diff
@@ -565,10 +567,20 @@ test("group-strategy", async () => {
         },
       ]);
 
+      const endpoints = {
```
[the remark this answers is not in the document]
Comment (narakhan) [reply lib/tests/group-strategy.test.ts:570]: I'd need to benchmark the `endpoints` object to give you a good answer here, but I'd expect similar issues to that of the `schemas` object.

Actually on this topic I had an idea that needs testing but might work. If Typescript applies the same optimisations to namespace modules as regular modules, then each section of the template (i.e. schemas, endpoints) could be contained by a namespace module, allowing for improved performance with basically the same interface being defined (also means the file grouped strategies can share an implementation, difference between them being one uses inlined namespace modules and the other separate files).

Tangentially related to the above, would you be alright if [this check](https://github.com/astahmer/openapi-zod-client/blob/3933722e0c6d7a7b7af50dd63d31aa5f91f666a1/lib/src/template-context.ts#LL98C29-L98C29) is yeeted. If endpoints are grouped under `default` (as is used for ungrouped endpoints in a grouped context, technically you could use the empty string for full backwards compatibility but I'd advise against it) in all situations rather than this special behaviour, then the `group.hbs` and `default.hbs` templates can reduced to *just* a single default template which greatly simplifies the libraries implementation.

Also what're your thoughts regarding making the default template just infer its schema types, haven't noticed any costs associated with this outside of a large line count, and it's almost always something you'll want to do as a consumer of the library anyway (as your applications types should flow from the API boundary directly) so it seems like it adds pointless boilerplate to not have it by default.

**Comment by narakhan** [comment]
@astahmer not going to have time to work on this till the weekend.

File: lib/tests/group-strategy.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (astahmer) [reply lib/tests/group-strategy.test.ts:570]: > Also what're your thoughts regarding making the default template just infer its schema types, haven't noticed any costs associated with this outside of a large line count, and it's almost always something you'll want to do as a consumer of the library anyway (as your applications types should flow from the API boundary directly) so it seems like it adds pointless boilerplate to not have it by default.

if that doesn't slow TS, I don't see any downside there, feel free to add this as well

**Comment by astahmer** [comment]
> @astahmer not going to have time to work on this till the weekend.

you've already done a lot ! take your time, there's no rush

File: lib/tests/group-strategy.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (narakhan) [reply lib/tests/group-strategy.test.ts:570]: > > In general if you're comfortable with some slight breaking changes (e.g. a default prefix by default) to the default template, then you can improve maintainability and performance. Avoided making changes like this before, but if you're fine with 'em I'll go ahead with them.
> 
> I am ! this PR already adds some breaking changes so, it feels a great timing if you want to add some more

Will do. You alright if I continue to overload this PR, rather than splitting it out into changesets? Would prefer to do so as many of the changes are dependant on each other, so it'd be less work on my end.

[the remark this answers is not in the document]
Comment (astahmer) [reply lib/tests/group-strategy.test.ts:570]: yes, please go for what's easiest for you to work with

File: lib/src/openApiToZod.ts (hunk 4 above)
Comment (narakhan) [reply lib/src/openApiToZod.ts:151]: Don't believe so. As a consumer of the library I'd expect a specification stating "there is an enum Foo with value Bar = 42" to allow the following code `Foo.Bar == 42`, that the enum happens to be a singleton doesn't really matter.

The non-labeled enum case is fine as it effectively doesn't change the exposed interface.

Comment (narakhan) [reply lib/src/openApiToZod.ts:151]: Eh — both NSwag and NJsonSchema (used by NSwag) have pretty terrible documentation, included a reference to the closest thing to a specification I could [find](https://github.com/RicoSuter/NJsonSchema/wiki/Enums#enum-names-and-descriptions).

**Comment by filipbekic01** [comment]
I ran into same performance issues. How is it going here?

**Comment by astahmer** [comment]
completely unrelated to this PR but could help with performance is another library I made that ditch the runtime validation library dependency or lets you use another one that has better IDE perfs
and it does use namespace as it was recommended here !

https://typed-openapi-astahmer.vercel.app/

**Comment by narakhan** [comment]
> I ran into same performance issues. How is it going here?

This PR alleviates the performance issues with IDEs — specifically type hinting — by a fair amount, but we were still observing poor performance for larger collections of endpoints. Though a relatively recent (around 2 months ago) update to Microsoft's Typescript language-server made it mostly bearable, they seem to have started caching the results of types on a larger variety of objects. So the first time you pull in anything from an openapi-zod generated module it takes around a minute (on my workplace's codebase), but afterwards hints are pretty much instant (until you close the IDE or the LSP yeets the cache).

So yeah if you're experiencing performance issues I'd advise bumping your Typescript LSP (it's probably Microsoft's), wont fix everything but makes it bearable.

---

Also @astahmer sorry about not finishing this PR, have had a combination of tight deadlines at work while being tasked with non-frontend areas of the codebase along with several major personal issues.

It's probably best just to merge the PR if you were okay with it, wont get around to extending the functionality anytime soon. [This](https://github.com/astahmer/openapi-zod-client/pull/134#discussion_r1184692045) outstanding issue is the only one unresolved, it was a while ago but if I recall correctly I tried your suggested solution but found it only obfuscated the typing issues and was still incorrect.

We've been using this PR at work for a couple of months now so it works on our machines™️, only annoying thing about it is the exporting template misses the generated types which aren't derived from the schema (to resolve Zod typing issues e.g. enums) as openapi-zod uses a different interfaces for these.

**Comment by astahmer** [comment]
regarding the missing types, this recent PR should help
https://github.com/astahmer/openapi-zod-client/pull/206

so you're saying this current PR is ready to merge ? minus the conflicts I mean

**Comment by narakhan** [comment]
> regarding the missing types, this recent PR should help #206
> 
> so you're saying this current PR is ready to merge ? minus the conflicts I mean

Was last time I checked, haven't been keeping up to date with the codebase though.

**Comment by astahmer** [comment]
got it. I rebased it here, will merge whenever someone takes the time to try and check that it's still working as intended
https://github.com/astahmer/openapi-zod-client/tree/feat/130-expose-predefined-templates-rebased
